<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.8"/>
<title>Gamgee: gamgee::Genotype Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
  $(document).ready(function() { searchBox.OnSelectItem(0); });
</script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td style="padding-left: 0.5em;">
   <div id="projectname">Gamgee
   </div>
   <div id="projectbrief">You miserable little maggot. I&#39;ll stove your head in!</div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.8 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li><a href="pages.html"><span>Related&#160;Pages</span></a></li>
      <li><a href="namespaces.html"><span>Namespaces</span></a></li>
      <li class="current"><a href="annotated.html"><span>Classes</span></a></li>
      <li><a href="files.html"><span>Files</span></a></li>
      <li><a href="examples.html"><span>Examples</span></a></li>
      <li>
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div id="navrow2" class="tabs2">
    <ul class="tablist">
      <li><a href="annotated.html"><span>Class&#160;List</span></a></li>
      <li><a href="classes.html"><span>Class&#160;Index</span></a></li>
      <li><a href="hierarchy.html"><span>Class&#160;Hierarchy</span></a></li>
      <li><a href="functions.html"><span>Class&#160;Members</span></a></li>
    </ul>
  </div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
<a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(0)"><span class="SelectionMark">&#160;</span>All</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(1)"><span class="SelectionMark">&#160;</span>Classes</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(2)"><span class="SelectionMark">&#160;</span>Namespaces</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(3)"><span class="SelectionMark">&#160;</span>Files</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(4)"><span class="SelectionMark">&#160;</span>Functions</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(5)"><span class="SelectionMark">&#160;</span>Variables</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(6)"><span class="SelectionMark">&#160;</span>Typedefs</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(7)"><span class="SelectionMark">&#160;</span>Enumerations</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(8)"><span class="SelectionMark">&#160;</span>Enumerator</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(9)"><span class="SelectionMark">&#160;</span>Friends</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(10)"><span class="SelectionMark">&#160;</span>Macros</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(11)"><span class="SelectionMark">&#160;</span>Pages</a></div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div id="nav-path" class="navpath">
  <ul>
<li class="navelem"><a class="el" href="namespacegamgee.html">gamgee</a></li><li class="navelem"><a class="el" href="classgamgee_1_1_genotype.html">Genotype</a></li>  </ul>
</div>
</div><!-- top -->
<div class="header">
  <div class="summary">
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="#pub-static-methods">Static Public Member Functions</a> &#124;
<a href="classgamgee_1_1_genotype-members.html">List of all members</a>  </div>
  <div class="headertitle">
<div class="title">gamgee::Genotype Class Reference</div>  </div>
</div><!--header-->
<div class="contents">

<p>Encodes a genotype.  
 <a href="classgamgee_1_1_genotype.html#details">More...</a></p>

<p><code>#include &lt;<a class="el" href="genotype_8h_source.html">genotype.h</a>&gt;</code></p>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr class="memitem:a3c548225a8dce6077aad6122906281f9"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classgamgee_1_1_genotype.html#a3c548225a8dce6077aad6122906281f9">Genotype</a> (const std::shared_ptr&lt; <a class="el" href="structbcf1__t.html">bcf1_t</a> &gt; &amp;body, const <a class="el" href="structbcf__fmt__t.html">bcf_fmt_t</a> *const format_ptr, const uint8_t *data_ptr)</td></tr>
<tr class="memdesc:a3c548225a8dce6077aad6122906281f9"><td class="mdescLeft">&#160;</td><td class="mdescRight">Constructs a genotype.  <a href="#a3c548225a8dce6077aad6122906281f9">More...</a><br /></td></tr>
<tr class="separator:a3c548225a8dce6077aad6122906281f9"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab3d22a0bf917d0fd42a162f41420b614"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classgamgee_1_1_genotype.html#ab3d22a0bf917d0fd42a162f41420b614">Genotype</a> (const <a class="el" href="classgamgee_1_1_genotype.html">Genotype</a> &amp;other)=delete</td></tr>
<tr class="memdesc:ab3d22a0bf917d0fd42a162f41420b614"><td class="mdescLeft">&#160;</td><td class="mdescRight">copying of the <a class="el" href="classgamgee_1_1_genotype.html" title="Encodes a genotype. ">Genotype</a> object is not allowed.  <a href="#ab3d22a0bf917d0fd42a162f41420b614">More...</a><br /></td></tr>
<tr class="separator:ab3d22a0bf917d0fd42a162f41420b614"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a476eb30243d02210058f6619e01fcd8c"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classgamgee_1_1_genotype.html">Genotype</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classgamgee_1_1_genotype.html#a476eb30243d02210058f6619e01fcd8c">operator=</a> (const <a class="el" href="classgamgee_1_1_genotype.html">Genotype</a> &amp;other)=delete</td></tr>
<tr class="memdesc:a476eb30243d02210058f6619e01fcd8c"><td class="mdescLeft">&#160;</td><td class="mdescRight">copying of the <a class="el" href="classgamgee_1_1_genotype.html" title="Encodes a genotype. ">Genotype</a> object is not allowed.  <a href="#a476eb30243d02210058f6619e01fcd8c">More...</a><br /></td></tr>
<tr class="separator:a476eb30243d02210058f6619e01fcd8c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a996ad9179e88898e44748f2f721463cd"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classgamgee_1_1_genotype.html#a996ad9179e88898e44748f2f721463cd">Genotype</a> (<a class="el" href="classgamgee_1_1_genotype.html">Genotype</a> &amp;&amp;other)=default</td></tr>
<tr class="memdesc:a996ad9179e88898e44748f2f721463cd"><td class="mdescLeft">&#160;</td><td class="mdescRight">Explicit default as recommended by many threads on stackoverflow.  <a href="#a996ad9179e88898e44748f2f721463cd">More...</a><br /></td></tr>
<tr class="separator:a996ad9179e88898e44748f2f721463cd"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab283d1c4cd484b750e1c1713afea0080"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classgamgee_1_1_genotype.html">Genotype</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classgamgee_1_1_genotype.html#ab283d1c4cd484b750e1c1713afea0080">operator=</a> (<a class="el" href="classgamgee_1_1_genotype.html">Genotype</a> &amp;&amp;other)=default</td></tr>
<tr class="memdesc:ab283d1c4cd484b750e1c1713afea0080"><td class="mdescLeft">&#160;</td><td class="mdescRight">Explicit default as recommended by many threads on stackoverflow.  <a href="#ab283d1c4cd484b750e1c1713afea0080">More...</a><br /></td></tr>
<tr class="separator:ab283d1c4cd484b750e1c1713afea0080"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7632e183ab71ca41630c088dc8de3737"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classgamgee_1_1_genotype.html#a7632e183ab71ca41630c088dc8de3737">~Genotype</a> ()=default</td></tr>
<tr class="memdesc:a7632e183ab71ca41630c088dc8de3737"><td class="mdescLeft">&#160;</td><td class="mdescRight">Explicit default as recommended by many threads on stackoverflow.  <a href="#a7632e183ab71ca41630c088dc8de3737">More...</a><br /></td></tr>
<tr class="separator:a7632e183ab71ca41630c088dc8de3737"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab76260c463d4d2cb082bbdab1c8e7e55"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classgamgee_1_1_genotype.html#ab76260c463d4d2cb082bbdab1c8e7e55">operator!=</a> (const <a class="el" href="classgamgee_1_1_genotype.html">Genotype</a> &amp;other) const </td></tr>
<tr class="memdesc:ab76260c463d4d2cb082bbdab1c8e7e55"><td class="mdescLeft">&#160;</td><td class="mdescRight">Checks if another genotype does not equal this genotype.  <a href="#ab76260c463d4d2cb082bbdab1c8e7e55">More...</a><br /></td></tr>
<tr class="separator:ab76260c463d4d2cb082bbdab1c8e7e55"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3a94d92eb5e550ddd0039cc46dcfb592"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classgamgee_1_1_genotype.html#a3a94d92eb5e550ddd0039cc46dcfb592">operator==</a> (const <a class="el" href="classgamgee_1_1_genotype.html">Genotype</a> &amp;other) const </td></tr>
<tr class="memdesc:a3a94d92eb5e550ddd0039cc46dcfb592"><td class="mdescLeft">&#160;</td><td class="mdescRight">Checks if another genotype equals this genotype.  <a href="#a3a94d92eb5e550ddd0039cc46dcfb592">More...</a><br /></td></tr>
<tr class="separator:a3a94d92eb5e550ddd0039cc46dcfb592"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a70d441d45b3cd89d5364d85d80e2c333"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classgamgee_1_1_genotype.html#a70d441d45b3cd89d5364d85d80e2c333">het</a> () const </td></tr>
<tr class="memdesc:a70d441d45b3cd89d5364d85d80e2c333"><td class="mdescLeft">&#160;</td><td class="mdescRight">Checks if this genotype vector is any type of heterozygous call.  <a href="#a70d441d45b3cd89d5364d85d80e2c333">More...</a><br /></td></tr>
<tr class="separator:a70d441d45b3cd89d5364d85d80e2c333"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a34e152f2cb8264f685c2722848308b1e"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classgamgee_1_1_genotype.html#a34e152f2cb8264f685c2722848308b1e">non_ref_het</a> () const </td></tr>
<tr class="memdesc:a34e152f2cb8264f685c2722848308b1e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Checks if this genotype vector is a heterozygous call and none of the alleles is the reference.  <a href="#a34e152f2cb8264f685c2722848308b1e">More...</a><br /></td></tr>
<tr class="separator:a34e152f2cb8264f685c2722848308b1e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a05e3002b133d74e407306a7aa7aeb8d5"><td class="memItemLeft" align="right" valign="top">uint32_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classgamgee_1_1_genotype.html#a05e3002b133d74e407306a7aa7aeb8d5">fast_diploid_key_generation</a> () const </td></tr>
<tr class="memdesc:a05e3002b133d74e407306a7aa7aeb8d5"><td class="mdescLeft">&#160;</td><td class="mdescRight">A bit encoding for the first two alleles.  <a href="#a05e3002b133d74e407306a7aa7aeb8d5">More...</a><br /></td></tr>
<tr class="separator:a05e3002b133d74e407306a7aa7aeb8d5"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a16825437893fa3876c424cf39e4cc8c1"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classgamgee_1_1_genotype.html#a16825437893fa3876c424cf39e4cc8c1">hom_var</a> () const </td></tr>
<tr class="memdesc:a16825437893fa3876c424cf39e4cc8c1"><td class="mdescLeft">&#160;</td><td class="mdescRight">Checks if this genotype vector is a homozygous call that is non-reference.  <a href="#a16825437893fa3876c424cf39e4cc8c1">More...</a><br /></td></tr>
<tr class="separator:a16825437893fa3876c424cf39e4cc8c1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae2c8e4b9deaa940ac5e4a7428a9310bd"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classgamgee_1_1_genotype.html#ae2c8e4b9deaa940ac5e4a7428a9310bd">hom_ref</a> () const </td></tr>
<tr class="memdesc:ae2c8e4b9deaa940ac5e4a7428a9310bd"><td class="mdescLeft">&#160;</td><td class="mdescRight">Checks if this genotype vector is a homozygous call that is reference.  <a href="#ae2c8e4b9deaa940ac5e4a7428a9310bd">More...</a><br /></td></tr>
<tr class="separator:ae2c8e4b9deaa940ac5e4a7428a9310bd"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a0fa76a9c8da2da98315a22dd81528143"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classgamgee_1_1_genotype.html#a0fa76a9c8da2da98315a22dd81528143">missing</a> () const </td></tr>
<tr class="memdesc:a0fa76a9c8da2da98315a22dd81528143"><td class="mdescLeft">&#160;</td><td class="mdescRight">Checks if all alleles are missing.  <a href="#a0fa76a9c8da2da98315a22dd81528143">More...</a><br /></td></tr>
<tr class="separator:a0fa76a9c8da2da98315a22dd81528143"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac332fe91cb41bda0bdeca1ecdd504162"><td class="memItemLeft" align="right" valign="top">std::vector&lt; std::string &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classgamgee_1_1_genotype.html#ac332fe91cb41bda0bdeca1ecdd504162">allele_strings</a> () const </td></tr>
<tr class="memdesc:ac332fe91cb41bda0bdeca1ecdd504162"><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns a vector with all the allele strings.  <a href="#ac332fe91cb41bda0bdeca1ecdd504162">More...</a><br /></td></tr>
<tr class="separator:ac332fe91cb41bda0bdeca1ecdd504162"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aaaf6fb73bcecc80d2cab17144765a9fc"><td class="memItemLeft" align="right" valign="top">std::vector&lt; int32_t &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classgamgee_1_1_genotype.html#aaaf6fb73bcecc80d2cab17144765a9fc">allele_keys</a> () const </td></tr>
<tr class="memdesc:aaaf6fb73bcecc80d2cab17144765a9fc"><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns a vector with all the allele keys.  <a href="#aaaf6fb73bcecc80d2cab17144765a9fc">More...</a><br /></td></tr>
<tr class="separator:aaaf6fb73bcecc80d2cab17144765a9fc"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa770dc86ecf6c0c162861d60b849db8a"><td class="memItemLeft" align="right" valign="top">std::string&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classgamgee_1_1_genotype.html#aa770dc86ecf6c0c162861d60b849db8a">allele_string</a> (const uint32_t index) const </td></tr>
<tr class="memdesc:aa770dc86ecf6c0c162861d60b849db8a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns the allele string at index.  <a href="#aa770dc86ecf6c0c162861d60b849db8a">More...</a><br /></td></tr>
<tr class="separator:aa770dc86ecf6c0c162861d60b849db8a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:abc3307ba31fda7dfec9653a703ff0716"><td class="memItemLeft" align="right" valign="top">int32_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classgamgee_1_1_genotype.html#abc3307ba31fda7dfec9653a703ff0716">allele_key</a> (const uint32_t index) const </td></tr>
<tr class="memdesc:abc3307ba31fda7dfec9653a703ff0716"><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns the allele key within this line.  <a href="#abc3307ba31fda7dfec9653a703ff0716">More...</a><br /></td></tr>
<tr class="separator:abc3307ba31fda7dfec9653a703ff0716"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a620ea021d732a1253c13ee80f75f3996"><td class="memItemLeft" align="right" valign="top">int32_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classgamgee_1_1_genotype.html#a620ea021d732a1253c13ee80f75f3996">operator[]</a> (const uint32_t index) const </td></tr>
<tr class="memdesc:a620ea021d732a1253c13ee80f75f3996"><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns the allele key within this line.  <a href="#a620ea021d732a1253c13ee80f75f3996">More...</a><br /></td></tr>
<tr class="separator:a620ea021d732a1253c13ee80f75f3996"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af5c9fc27656bda0a716982f073d5f794"><td class="memItemLeft" align="right" valign="top">uint32_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classgamgee_1_1_genotype.html#af5c9fc27656bda0a716982f073d5f794">size</a> () const </td></tr>
<tr class="memdesc:af5c9fc27656bda0a716982f073d5f794"><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns the number of alleles.  <a href="#af5c9fc27656bda0a716982f073d5f794">More...</a><br /></td></tr>
<tr class="separator:af5c9fc27656bda0a716982f073d5f794"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3826cabaacce33beb30e8a1626cc8fee"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classgamgee_1_1_genotype.html#a3826cabaacce33beb30e8a1626cc8fee">snp</a> (const <a class="el" href="namespacegamgee.html#a59662b9ac19e4dc28f4ea79e7566a348">AlleleMask</a> &amp;mask) const </td></tr>
<tr class="memdesc:a3826cabaacce33beb30e8a1626cc8fee"><td class="mdescLeft">&#160;</td><td class="mdescRight">whether or not this genotype represents a snp  <a href="#a3826cabaacce33beb30e8a1626cc8fee">More...</a><br /></td></tr>
<tr class="separator:a3826cabaacce33beb30e8a1626cc8fee"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af9d1aa1e199d3ec48c22b1e95c5d0f2b"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classgamgee_1_1_genotype.html#af9d1aa1e199d3ec48c22b1e95c5d0f2b">insertion</a> (const <a class="el" href="namespacegamgee.html#a59662b9ac19e4dc28f4ea79e7566a348">AlleleMask</a> &amp;mask) const </td></tr>
<tr class="memdesc:af9d1aa1e199d3ec48c22b1e95c5d0f2b"><td class="mdescLeft">&#160;</td><td class="mdescRight">whether or not this genotype represents an insertion  <a href="#af9d1aa1e199d3ec48c22b1e95c5d0f2b">More...</a><br /></td></tr>
<tr class="separator:af9d1aa1e199d3ec48c22b1e95c5d0f2b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7190bf6a07c7087e1cc85b9ce3a6ec6b"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classgamgee_1_1_genotype.html#a7190bf6a07c7087e1cc85b9ce3a6ec6b">deletion</a> (const <a class="el" href="namespacegamgee.html#a59662b9ac19e4dc28f4ea79e7566a348">AlleleMask</a> &amp;mask) const </td></tr>
<tr class="memdesc:a7190bf6a07c7087e1cc85b9ce3a6ec6b"><td class="mdescLeft">&#160;</td><td class="mdescRight">whether or not this genotype represents an deletion  <a href="#a7190bf6a07c7087e1cc85b9ce3a6ec6b">More...</a><br /></td></tr>
<tr class="separator:a7190bf6a07c7087e1cc85b9ce3a6ec6b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7eda29ef8b52af32d7032889aed5c793"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classgamgee_1_1_genotype.html#a7eda29ef8b52af32d7032889aed5c793">indel</a> (const <a class="el" href="namespacegamgee.html#a59662b9ac19e4dc28f4ea79e7566a348">AlleleMask</a> &amp;mask) const </td></tr>
<tr class="memdesc:a7eda29ef8b52af32d7032889aed5c793"><td class="mdescLeft">&#160;</td><td class="mdescRight">whether or not this genotype represents an insertion or deletion  <a href="#a7eda29ef8b52af32d7032889aed5c793">More...</a><br /></td></tr>
<tr class="separator:a7eda29ef8b52af32d7032889aed5c793"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a9dbd0e337861f917b6a0c83ae222b1cf"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classgamgee_1_1_genotype.html#a9dbd0e337861f917b6a0c83ae222b1cf">biallelic</a> () const </td></tr>
<tr class="memdesc:a9dbd0e337861f917b6a0c83ae222b1cf"><td class="mdescLeft">&#160;</td><td class="mdescRight">whether or not this genotype has <em>at most</em> one alternate allele  <a href="#a9dbd0e337861f917b6a0c83ae222b1cf">More...</a><br /></td></tr>
<tr class="separator:a9dbd0e337861f917b6a0c83ae222b1cf"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a440f136f7b9f2a05ad0fc8223159c193"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classgamgee_1_1_genotype.html#a440f136f7b9f2a05ad0fc8223159c193">complex</a> () const </td></tr>
<tr class="memdesc:a440f136f7b9f2a05ad0fc8223159c193"><td class="mdescLeft">&#160;</td><td class="mdescRight">literally the negation of biallelic(mask)  <a href="#a440f136f7b9f2a05ad0fc8223159c193">More...</a><br /></td></tr>
<tr class="separator:a440f136f7b9f2a05ad0fc8223159c193"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a80f98ec5ab562872f20e18dccd85063b"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classgamgee_1_1_genotype.html#a80f98ec5ab562872f20e18dccd85063b">mixed</a> () const </td></tr>
<tr class="memdesc:a80f98ec5ab562872f20e18dccd85063b"><td class="mdescLeft">&#160;</td><td class="mdescRight">identifies variants with two different types of alleles  <a href="#a80f98ec5ab562872f20e18dccd85063b">More...</a><br /></td></tr>
<tr class="separator:a80f98ec5ab562872f20e18dccd85063b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a0bc8bb03185dd4423b65feb881e97f36"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classgamgee_1_1_genotype.html#a0bc8bb03185dd4423b65feb881e97f36">variant</a> () const </td></tr>
<tr class="separator:a0bc8bb03185dd4423b65feb881e97f36"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-static-methods"></a>
Static Public Member Functions</h2></td></tr>
<tr class="memitem:a014bc9217f7c78d56ae017da672ceb2a"><td class="memItemLeft" align="right" valign="top">static void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classgamgee_1_1_genotype.html#a014bc9217f7c78d56ae017da672ceb2a">encode_genotype</a> (std::vector&lt; int32_t &gt; &amp;alleles)</td></tr>
<tr class="memdesc:a014bc9217f7c78d56ae017da672ceb2a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Converts a vector of allele indices representing a genotype into BCF-encoded format suitable for passing to <a class="el" href="classgamgee_1_1_variant_builder.html#a24a3d88724ad637491b14d89e320dd4e">VariantBuilder::set_genotypes()</a>. No phasing is added.  <a href="#a014bc9217f7c78d56ae017da672ceb2a">More...</a><br /></td></tr>
<tr class="separator:a014bc9217f7c78d56ae017da672ceb2a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab68c9f3be8754d9aef2a6190f9ea1acf"><td class="memItemLeft" align="right" valign="top">static void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classgamgee_1_1_genotype.html#ab68c9f3be8754d9aef2a6190f9ea1acf">encode_genotype</a> (std::vector&lt; int32_t &gt; &amp;alleles, bool phase_all_alleles)</td></tr>
<tr class="memdesc:ab68c9f3be8754d9aef2a6190f9ea1acf"><td class="mdescLeft">&#160;</td><td class="mdescRight">Converts a vector of allele indices representing a genotype into BCF-encoded format suitable for passing to <a class="el" href="classgamgee_1_1_variant_builder.html#a24a3d88724ad637491b14d89e320dd4e">VariantBuilder::set_genotypes()</a>, and also allows you to phase all alleles.  <a href="#ab68c9f3be8754d9aef2a6190f9ea1acf">More...</a><br /></td></tr>
<tr class="separator:ab68c9f3be8754d9aef2a6190f9ea1acf"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac62b7ad602284c04a4f6e159471fb325"><td class="memItemLeft" align="right" valign="top">static void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classgamgee_1_1_genotype.html#ac62b7ad602284c04a4f6e159471fb325">encode_genotypes</a> (std::vector&lt; std::vector&lt; int32_t &gt;&gt; &amp;multiple_genotypes)</td></tr>
<tr class="memdesc:ac62b7ad602284c04a4f6e159471fb325"><td class="mdescLeft">&#160;</td><td class="mdescRight">Converts multiple vectors of allele indices representing genotypes into BCF_encoded format suitable for passing to <a class="el" href="classgamgee_1_1_variant_builder.html#a24a3d88724ad637491b14d89e320dd4e">VariantBuilder::set_genotypes()</a>. No phasing is added.  <a href="#ac62b7ad602284c04a4f6e159471fb325">More...</a><br /></td></tr>
<tr class="separator:ac62b7ad602284c04a4f6e159471fb325"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><p>Encodes a genotype. </p>
</div><h2 class="groupheader">Constructor &amp; Destructor Documentation</h2>
<a class="anchor" id="a3c548225a8dce6077aad6122906281f9"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">gamgee::Genotype::Genotype </td>
          <td>(</td>
          <td class="paramtype">const std::shared_ptr&lt; <a class="el" href="structbcf1__t.html">bcf1_t</a> &gt; &amp;&#160;</td>
          <td class="paramname"><em>body</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structbcf__fmt__t.html">bcf_fmt_t</a> *const&#160;</td>
          <td class="paramname"><em>format_ptr</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const uint8_t *&#160;</td>
          <td class="paramname"><em>data_ptr</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Constructs a genotype. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">body</td><td>The shared memory variant "line" from a vcf, or bcf. </td></tr>
    <tr><td class="paramname">format_ptr</td><td>The GT field from the line. </td></tr>
    <tr><td class="paramname">data_ptr</td><td>The GT for this sample.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section note"><dt>Note</dt><dd>Most of the patterns here are directly copied or adapted from htslib. </dd></dl>

</div>
</div>
<a class="anchor" id="ab3d22a0bf917d0fd42a162f41420b614"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">gamgee::Genotype::Genotype </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="classgamgee_1_1_genotype.html">Genotype</a> &amp;&#160;</td>
          <td class="paramname"><em>other</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">delete</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>copying of the <a class="el" href="classgamgee_1_1_genotype.html" title="Encodes a genotype. ">Genotype</a> object is not allowed. </p>

</div>
</div>
<a class="anchor" id="a996ad9179e88898e44748f2f721463cd"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">gamgee::Genotype::Genotype </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="classgamgee_1_1_genotype.html">Genotype</a> &amp;&amp;&#160;</td>
          <td class="paramname"><em>other</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">default</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Explicit default as recommended by many threads on stackoverflow. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">other</td><td>Other genotype. </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a class="anchor" id="a7632e183ab71ca41630c088dc8de3737"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">gamgee::Genotype::~Genotype </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">default</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Explicit default as recommended by many threads on stackoverflow. </p>

</div>
</div>
<h2 class="groupheader">Member Function Documentation</h2>
<a class="anchor" id="abc3307ba31fda7dfec9653a703ff0716"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int32_t gamgee::Genotype::allele_key </td>
          <td>(</td>
          <td class="paramtype">const uint32_t&#160;</td>
          <td class="paramname"><em>index</em></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Returns the allele key within this line. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">index</td><td>Zero based allele index </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>the allele key within this line </dd></dl>

</div>
</div>
<a class="anchor" id="aaaf6fb73bcecc80d2cab17144765a9fc"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">std::vector&lt; int32_t &gt; gamgee::Genotype::allele_keys </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Returns a vector with all the allele keys. </p>
<dl class="section return"><dt>Returns</dt><dd>a vector with all the allele keys </dd></dl>

</div>
</div>
<a class="anchor" id="aa770dc86ecf6c0c162861d60b849db8a"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">std::string gamgee::Genotype::allele_string </td>
          <td>(</td>
          <td class="paramtype">const uint32_t&#160;</td>
          <td class="paramname"><em>index</em></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Returns the allele string at index. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">index</td><td>Zero based allele index </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>the allele string </dd></dl>

</div>
</div>
<a class="anchor" id="ac332fe91cb41bda0bdeca1ecdd504162"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">std::vector&lt; std::string &gt; gamgee::Genotype::allele_strings </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Returns a vector with all the allele strings. </p>
<dl class="section return"><dt>Returns</dt><dd>a vector with all the allele strings </dd></dl>

</div>
</div>
<a class="anchor" id="a9dbd0e337861f917b6a0c83ae222b1cf"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool gamgee::Genotype::biallelic </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>whether or not this genotype has <em>at most</em> one alternate allele </p>
<p>This function will check whether this is a simple heterozygous or homozygous site where both alleles are either the same, or reference and one alternate allele. No two different alleles would pass.</p>
<dl class="section warning"><dt>Warning</dt><dd>complexity is O(n) in the number of alleles. If you are using many of these convenience functions (<a class="el" href="classgamgee_1_1_genotype.html#a3826cabaacce33beb30e8a1626cc8fee" title="whether or not this genotype represents a snp ">snp()</a>, <a class="el" href="classgamgee_1_1_genotype.html#af9d1aa1e199d3ec48c22b1e95c5d0f2b" title="whether or not this genotype represents an insertion ">insertion()</a>, <a class="el" href="classgamgee_1_1_genotype.html#a7190bf6a07c7087e1cc85b9ce3a6ec6b" title="whether or not this genotype represents an deletion ">deletion()</a>, <a class="el" href="classgamgee_1_1_genotype.html#a7eda29ef8b52af32d7032889aed5c793" title="whether or not this genotype represents an insertion or deletion ">indel()</a>, <a class="el" href="classgamgee_1_1_genotype.html#a440f136f7b9f2a05ad0fc8223159c193" title="literally the negation of biallelic(mask) ">complex()</a>,...), you will be better off implementing one loop that makes all the checks in one pass instead of calling many O(n) functions.</dd></dl>
<dl class="section return"><dt>Returns</dt><dd>whether or not all alleles are either the same, or there is one alt allele mixed with reference alleles. </dd></dl>

</div>
</div>
<a class="anchor" id="a440f136f7b9f2a05ad0fc8223159c193"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">bool gamgee::Genotype::complex </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>literally the negation of biallelic(mask) </p>
<dl class="section warning"><dt>Warning</dt><dd>complexity is O(n) in the number of alleles. If you are using many of these convenience functions (<a class="el" href="classgamgee_1_1_genotype.html#a3826cabaacce33beb30e8a1626cc8fee" title="whether or not this genotype represents a snp ">snp()</a>, <a class="el" href="classgamgee_1_1_genotype.html#af9d1aa1e199d3ec48c22b1e95c5d0f2b" title="whether or not this genotype represents an insertion ">insertion()</a>, <a class="el" href="classgamgee_1_1_genotype.html#a7190bf6a07c7087e1cc85b9ce3a6ec6b" title="whether or not this genotype represents an deletion ">deletion()</a>, <a class="el" href="classgamgee_1_1_genotype.html#a7eda29ef8b52af32d7032889aed5c793" title="whether or not this genotype represents an insertion or deletion ">indel()</a>, <a class="el" href="classgamgee_1_1_genotype.html#a440f136f7b9f2a05ad0fc8223159c193" title="literally the negation of biallelic(mask) ">complex()</a>,...), you will be better off implementing one loop that makes all the checks in one pass instead of calling many O(n) functions.</dd></dl>
<dl class="section return"><dt>Returns</dt><dd>whether or not there are more than one alt allele in this record </dd></dl>

</div>
</div>
<a class="anchor" id="a7190bf6a07c7087e1cc85b9ce3a6ec6b"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool gamgee::Genotype::deletion </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="namespacegamgee.html#a59662b9ac19e4dc28f4ea79e7566a348">AlleleMask</a> &amp;&#160;</td>
          <td class="paramname"><em>mask</em></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>whether or not this genotype represents an deletion </p>
<p>This function will check that at least one deletion exists and that no other type of allele (snp or insertion) is present. Multiple deletions will return true.</p>
<dl class="section warning"><dt>Warning</dt><dd>complexity is O(n) in the number of alleles. If you are using many of these convenience functions (<a class="el" href="classgamgee_1_1_genotype.html#a3826cabaacce33beb30e8a1626cc8fee" title="whether or not this genotype represents a snp ">snp()</a>, <a class="el" href="classgamgee_1_1_genotype.html#af9d1aa1e199d3ec48c22b1e95c5d0f2b" title="whether or not this genotype represents an insertion ">insertion()</a>, <a class="el" href="classgamgee_1_1_genotype.html#a7190bf6a07c7087e1cc85b9ce3a6ec6b" title="whether or not this genotype represents an deletion ">deletion()</a>, <a class="el" href="classgamgee_1_1_genotype.html#a7eda29ef8b52af32d7032889aed5c793" title="whether or not this genotype represents an insertion or deletion ">indel()</a>, <a class="el" href="classgamgee_1_1_genotype.html#a440f136f7b9f2a05ad0fc8223159c193" title="literally the negation of biallelic(mask) ">complex()</a>,...), you will be better off implementing one loop that makes all the checks in one pass instead of calling many O(n) functions.</dd></dl>
<dl class="section return"><dt>Returns</dt><dd>whether or not there is at least one deletion in this genotype and nothing else but deletions and reference alleles </dd></dl>

</div>
</div>
<a class="anchor" id="a014bc9217f7c78d56ae017da672ceb2a"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static void gamgee::Genotype::encode_genotype </td>
          <td>(</td>
          <td class="paramtype">std::vector&lt; int32_t &gt; &amp;&#160;</td>
          <td class="paramname"><em>alleles</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Converts a vector of allele indices representing a genotype into BCF-encoded format suitable for passing to <a class="el" href="classgamgee_1_1_variant_builder.html#a24a3d88724ad637491b14d89e320dd4e">VariantBuilder::set_genotypes()</a>. No phasing is added. </p>
<p>Example: if you want to encode the genotype 0/1, create a vector with {0, 1} and then pass it to this function </p>

</div>
</div>
<a class="anchor" id="ab68c9f3be8754d9aef2a6190f9ea1acf"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static void gamgee::Genotype::encode_genotype </td>
          <td>(</td>
          <td class="paramtype">std::vector&lt; int32_t &gt; &amp;&#160;</td>
          <td class="paramname"><em>alleles</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">bool&#160;</td>
          <td class="paramname"><em>phase_all_alleles</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Converts a vector of allele indices representing a genotype into BCF-encoded format suitable for passing to <a class="el" href="classgamgee_1_1_variant_builder.html#a24a3d88724ad637491b14d89e320dd4e">VariantBuilder::set_genotypes()</a>, and also allows you to phase all alleles. </p>
<p>Example: if you want to encode the genotype 0|1, create a vector with {0, 1} and then pass it to this function with phase_all_alleles set to true </p>

</div>
</div>
<a class="anchor" id="ac62b7ad602284c04a4f6e159471fb325"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static void gamgee::Genotype::encode_genotypes </td>
          <td>(</td>
          <td class="paramtype">std::vector&lt; std::vector&lt; int32_t &gt;&gt; &amp;&#160;</td>
          <td class="paramname"><em>multiple_genotypes</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Converts multiple vectors of allele indices representing genotypes into BCF_encoded format suitable for passing to <a class="el" href="classgamgee_1_1_variant_builder.html#a24a3d88724ad637491b14d89e320dd4e">VariantBuilder::set_genotypes()</a>. No phasing is added. </p>
<p>Example: if you want to encode the genotypes 0/1 and 1/1, create a vector with { {0, 1}, {1, 1} } and pass it to this function </p>

</div>
</div>
<a class="anchor" id="a05e3002b133d74e407306a7aa7aeb8d5"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">uint32_t gamgee::Genotype::fast_diploid_key_generation </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>A bit encoding for the first two alleles. </p>
<dl class="section return"><dt>Returns</dt><dd>A bit encoding for the first two alleles. </dd></dl>
<dl class="section note"><dt>Note</dt><dd>only for diploids, returns false otherwise. </dd></dl>

</div>
</div>
<a class="anchor" id="a70d441d45b3cd89d5364d85d80e2c333"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool gamgee::Genotype::het </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Checks if this genotype vector is any type of heterozygous call. </p>
<dl class="section return"><dt>Returns</dt><dd>True if this GT is a het. </dd></dl>
<dl class="section note"><dt>Note</dt><dd>only for diploids. </dd></dl>

</div>
</div>
<a class="anchor" id="ae2c8e4b9deaa940ac5e4a7428a9310bd"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool gamgee::Genotype::hom_ref </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Checks if this genotype vector is a homozygous call that is reference. </p>
<dl class="section return"><dt>Returns</dt><dd>True if this GT is a hom ref. </dd></dl>

</div>
</div>
<a class="anchor" id="a16825437893fa3876c424cf39e4cc8c1"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool gamgee::Genotype::hom_var </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Checks if this genotype vector is a homozygous call that is non-reference. </p>
<dl class="section return"><dt>Returns</dt><dd>True if this GT is a hom var. </dd></dl>

</div>
</div>
<a class="anchor" id="a7eda29ef8b52af32d7032889aed5c793"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool gamgee::Genotype::indel </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="namespacegamgee.html#a59662b9ac19e4dc28f4ea79e7566a348">AlleleMask</a> &amp;&#160;</td>
          <td class="paramname"><em>mask</em></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>whether or not this genotype represents an insertion or deletion </p>
<p>This function will check that at least one insertion or deletion exists and that no other type of allele (snp or insertion) is present. Multiple deletions will return true.</p>
<dl class="section note"><dt>Note</dt><dd>this is not the same as insertion(mask) || deletion(mask) because it will also tolerate sites with insertions and deletions, while both other functions would return false to such a site.</dd></dl>
<dl class="section warning"><dt>Warning</dt><dd>complexity is O(n) in the number of alleles. If you are using many of these convenience functions (<a class="el" href="classgamgee_1_1_genotype.html#a3826cabaacce33beb30e8a1626cc8fee" title="whether or not this genotype represents a snp ">snp()</a>, <a class="el" href="classgamgee_1_1_genotype.html#af9d1aa1e199d3ec48c22b1e95c5d0f2b" title="whether or not this genotype represents an insertion ">insertion()</a>, <a class="el" href="classgamgee_1_1_genotype.html#a7190bf6a07c7087e1cc85b9ce3a6ec6b" title="whether or not this genotype represents an deletion ">deletion()</a>, <a class="el" href="classgamgee_1_1_genotype.html#a7eda29ef8b52af32d7032889aed5c793" title="whether or not this genotype represents an insertion or deletion ">indel()</a>, <a class="el" href="classgamgee_1_1_genotype.html#a440f136f7b9f2a05ad0fc8223159c193" title="literally the negation of biallelic(mask) ">complex()</a>,...), you will be better off implementing one loop that makes all the checks in one pass instead of calling many O(n) functions.</dd></dl>
<dl class="section return"><dt>Returns</dt><dd>whether or not there is at least one insertion or deletion in this genotype and nothing else but insertions, deletions and reference alleles </dd></dl>

</div>
</div>
<a class="anchor" id="af9d1aa1e199d3ec48c22b1e95c5d0f2b"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool gamgee::Genotype::insertion </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="namespacegamgee.html#a59662b9ac19e4dc28f4ea79e7566a348">AlleleMask</a> &amp;&#160;</td>
          <td class="paramname"><em>mask</em></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>whether or not this genotype represents an insertion </p>
<p>This function will check that at least one insertion exists and that no other type of allele (snp or deletion) is present. Multiple insertions will return true.</p>
<dl class="section warning"><dt>Warning</dt><dd>complexity is O(n) in the number of alleles. If you are using many of these convenience functions (<a class="el" href="classgamgee_1_1_genotype.html#a3826cabaacce33beb30e8a1626cc8fee" title="whether or not this genotype represents a snp ">snp()</a>, <a class="el" href="classgamgee_1_1_genotype.html#af9d1aa1e199d3ec48c22b1e95c5d0f2b" title="whether or not this genotype represents an insertion ">insertion()</a>, <a class="el" href="classgamgee_1_1_genotype.html#a7190bf6a07c7087e1cc85b9ce3a6ec6b" title="whether or not this genotype represents an deletion ">deletion()</a>, <a class="el" href="classgamgee_1_1_genotype.html#a7eda29ef8b52af32d7032889aed5c793" title="whether or not this genotype represents an insertion or deletion ">indel()</a>, <a class="el" href="classgamgee_1_1_genotype.html#a440f136f7b9f2a05ad0fc8223159c193" title="literally the negation of biallelic(mask) ">complex()</a>,...), you will be better off implementing one loop that makes all the checks in one pass instead of calling many O(n) functions.</dd></dl>
<dl class="section return"><dt>Returns</dt><dd>whether or not there is at least one insertion in this genotype and nothing else but insertions and reference alleles </dd></dl>

</div>
</div>
<a class="anchor" id="a0fa76a9c8da2da98315a22dd81528143"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool gamgee::Genotype::missing </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Checks if all alleles are missing. </p>
<dl class="section return"><dt>Returns</dt><dd>True if all alleles are missing. </dd></dl>
<dl class="section warning"><dt>Warning</dt><dd>Missing GT fields are untested. </dd></dl>

</div>
</div>
<a class="anchor" id="a80f98ec5ab562872f20e18dccd85063b"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool gamgee::Genotype::mixed </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>identifies variants with two different types of alleles </p>
<dl class="section warning"><dt>Warning</dt><dd>complexity is O(n) in the number of alleles. If you are using many of these convenience functions (<a class="el" href="classgamgee_1_1_genotype.html#a3826cabaacce33beb30e8a1626cc8fee" title="whether or not this genotype represents a snp ">snp()</a>, <a class="el" href="classgamgee_1_1_genotype.html#af9d1aa1e199d3ec48c22b1e95c5d0f2b" title="whether or not this genotype represents an insertion ">insertion()</a>, <a class="el" href="classgamgee_1_1_genotype.html#a7190bf6a07c7087e1cc85b9ce3a6ec6b" title="whether or not this genotype represents an deletion ">deletion()</a>, <a class="el" href="classgamgee_1_1_genotype.html#a7eda29ef8b52af32d7032889aed5c793" title="whether or not this genotype represents an insertion or deletion ">indel()</a>, <a class="el" href="classgamgee_1_1_genotype.html#a440f136f7b9f2a05ad0fc8223159c193" title="literally the negation of biallelic(mask) ">complex()</a>,...), you will be better off implementing one loop that makes all the checks in one pass instead of calling many O(n) functions.</dd></dl>
<dl class="section return"><dt>Returns</dt><dd>whether or not there are more than one types of alt allele in this record </dd></dl>

</div>
</div>
<a class="anchor" id="a34e152f2cb8264f685c2722848308b1e"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool gamgee::Genotype::non_ref_het </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Checks if this genotype vector is a heterozygous call and none of the alleles is the reference. </p>
<dl class="section return"><dt>Returns</dt><dd>True if this GT is a het and none of the alleles is the reference. </dd></dl>
<dl class="section note"><dt>Note</dt><dd>only for diploids. </dd></dl>

</div>
</div>
<a class="anchor" id="ab76260c463d4d2cb082bbdab1c8e7e55"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool gamgee::Genotype::operator!= </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="classgamgee_1_1_genotype.html">Genotype</a> &amp;&#160;</td>
          <td class="paramname"><em>other</em></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Checks if another genotype does not equal this genotype. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">other</td><td>The other genotype to compare. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>False if other genotype equals this genotype. </dd></dl>

</div>
</div>
<a class="anchor" id="a476eb30243d02210058f6619e01fcd8c"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classgamgee_1_1_genotype.html">Genotype</a>&amp; gamgee::Genotype::operator= </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="classgamgee_1_1_genotype.html">Genotype</a> &amp;&#160;</td>
          <td class="paramname"><em>other</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">delete</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>copying of the <a class="el" href="classgamgee_1_1_genotype.html" title="Encodes a genotype. ">Genotype</a> object is not allowed. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">other</td><td>Other genotype. </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a class="anchor" id="ab283d1c4cd484b750e1c1713afea0080"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classgamgee_1_1_genotype.html">Genotype</a>&amp; gamgee::Genotype::operator= </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="classgamgee_1_1_genotype.html">Genotype</a> &amp;&amp;&#160;</td>
          <td class="paramname"><em>other</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">default</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Explicit default as recommended by many threads on stackoverflow. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">other</td><td>Other genotype. </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a class="anchor" id="a3a94d92eb5e550ddd0039cc46dcfb592"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool gamgee::Genotype::operator== </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="classgamgee_1_1_genotype.html">Genotype</a> &amp;&#160;</td>
          <td class="paramname"><em>other</em></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Checks if another genotype equals this genotype. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">other</td><td>The other genotype to compare. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>True if other genotype equals this genotype. </dd></dl>
<dl class="section note"><dt>Note</dt><dd>No string comparison is done. This operation is supposed to be fast. </dd></dl>

</div>
</div>
<a class="anchor" id="a620ea021d732a1253c13ee80f75f3996"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int32_t gamgee::Genotype::operator[] </td>
          <td>(</td>
          <td class="paramtype">const uint32_t&#160;</td>
          <td class="paramname"><em>index</em></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Returns the allele key within this line. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">index</td><td>Zero based allele index </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>the allele key within this line </dd></dl>

</div>
</div>
<a class="anchor" id="af5c9fc27656bda0a716982f073d5f794"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">uint32_t gamgee::Genotype::size </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Returns the number of alleles. </p>
<dl class="section return"><dt>Returns</dt><dd>the number of alleles </dd></dl>

</div>
</div>
<a class="anchor" id="a3826cabaacce33beb30e8a1626cc8fee"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool gamgee::Genotype::snp </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="namespacegamgee.html#a59662b9ac19e4dc28f4ea79e7566a348">AlleleMask</a> &amp;&#160;</td>
          <td class="paramname"><em>mask</em></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>whether or not this genotype represents a snp </p>
<p>In the true sense of the word, single nucleotide polymorphism restricts the number of loci (nucleotides) that are polymorphic, but not the number of potentially different alleles it may have (as long as they are all single nucleotide polymorphisms). This function will check that at least one of these SNPs exists and that no other type of allele (insertion or deletion) is present. Multiple SNPs will return true.</p>
<dl class="section warning"><dt>Warning</dt><dd>complexity is O(n) in the number of alleles. If you are using many of these convenience functions (<a class="el" href="classgamgee_1_1_genotype.html#a3826cabaacce33beb30e8a1626cc8fee" title="whether or not this genotype represents a snp ">snp()</a>, <a class="el" href="classgamgee_1_1_genotype.html#af9d1aa1e199d3ec48c22b1e95c5d0f2b" title="whether or not this genotype represents an insertion ">insertion()</a>, <a class="el" href="classgamgee_1_1_genotype.html#a7190bf6a07c7087e1cc85b9ce3a6ec6b" title="whether or not this genotype represents an deletion ">deletion()</a>, <a class="el" href="classgamgee_1_1_genotype.html#a7eda29ef8b52af32d7032889aed5c793" title="whether or not this genotype represents an insertion or deletion ">indel()</a>, <a class="el" href="classgamgee_1_1_genotype.html#a440f136f7b9f2a05ad0fc8223159c193" title="literally the negation of biallelic(mask) ">complex()</a>,...), you will be better off implementing one loop that makes all the checks in one pass instead of calling many O(n) functions.</dd></dl>
<dl class="section return"><dt>Returns</dt><dd>whether or not there is at least one snp in this genotype and nothing else but snps and reference alleles </dd></dl>

</div>
</div>
<a class="anchor" id="a0bc8bb03185dd4423b65feb881e97f36"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">bool gamgee::Genotype::variant </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<hr/>The documentation for this class was generated from the following files:<ul>
<li>gamgee/<a class="el" href="genotype_8h_source.html">genotype.h</a></li>
<li>gamgee/<a class="el" href="genotype_8cpp.html">genotype.cpp</a></li>
</ul>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated on Tue Nov 4 2014 20:03:46 for Gamgee by &#160;<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/>
</a> 1.8.8
</small></address>
</body>
</html>
